[pt] Translate /pt/docs/languages/go/instrumentation (#5380)
open-telemetry/opentelemetry.io: content/pt/docs/languages/go/instrumentation.md

```go
var myKey = attribute.Key("meuAtributoLegal")
span.SetAttributes(myKey.String("um valor em texto"))
```

#### Atributos Semânticos {#semantic-attributes}

Os Atributos Semânticos são atributos definidos pela [Especificação do
OpenTelemetry][opentelemetry specification] para fornecer um conjunto comum de
chaves de atributos entre várias linguagens, frameworks e ambientes de execução.
Esses atributos representam conceitos como métodos HTTP, códigos de estado, user
agents e outros. Estes atributos estão disponíveis no pacote
`go.opentelemetry.io/otel/semconv/v1.26.0`.

Para mais detalhes, consulte as [Convenções Semânticas de
Rastros][trace semantic conventions].

### Eventos {#events}

Um evento é uma mensagem legível para humanos em um trecho que representa "algo
acontecendo" durante a sua duração. Por exemplo, imagine uma função que requer
acesso exclusivo a um recurso que está sob um mutex. Um evento poderia ser
criado em dois pontos: um quando tentamos obter acesso ao recurso e outro quando
adquirimos o mutex.

```go
span.AddEvent("Obtendo o lock")
mutex.Lock()
span.AddEvent("Lock efetuado, realizando o trabalho...")
// do stuff

> span.AddEvent("Realizando unlocking")

Reviewer: ```suggestion
span.AddEvent("Realizando o desbloqueio")
```

mutex.Unlock()
```

> Uma característica útil dos eventos é que seus timestamps são exibidos como

Reviewer: ```suggestion
Uma característica útil dos eventos é que seus registro de data e hora são exibidos como
```

> offsets a partir do início do trecho, permitindo ver facilmente quanto tempo se

Reviewer: ```suggestion
deslocamentos a partir do início do trecho, permitindo ver facilmente quanto tempo se
```

Reviewer: Não tenho certeza se deslocamentos é a melhor palavra para traduzir offsets.

Author: Ops, acabei aceitando esse aqui, mas concordo com `intervalos`!

> offsets a partir do início do trecho, permitindo ver facilmente quanto tempo se

Reviewer: ```suggestion
intervalos a partir do início do trecho, permitindo ver facilmente quanto tempo se
```

```go
span.AddEvent("Espera cancelada devido sinal externo", trace.WithAttributes(attribute.Int("pid", 4328), attribute.String("signal", "SIGHUP")))
```

### Definir status do Trecho {#set-span-status}

{{% pt/docs/languages/span-status-preamble %}}

```go
import (
	// ...
	"go.opentelemetry.io/otel/codes"
	// ...
)

// ...

result, err := operacaoQueDeveriaFalhar()
if err != nil {
	span.SetStatus(codes.Error, "operacaoQueDeveriaFalhar falhou")
}
```

### Capturar erros {#record-errors}

Caso você tenha uma operação que falhou e deseja capturar o erro produzido, você
pode registrar este erro.

```go
import (
	// ...
	"go.opentelemetry.io/otel/codes"
	// ...
)

// ...

result, err := operacaoQueDeveriaFalhar()
if err != nil {
	span.SetStatus(codes.Error, "operacaoQueDeveriaFalhar falhou")
	span.RecordError(err)
}
```

É altamente recomendável que você também defina o estado de um trecho como
`Error` ao utilizar `RecordError`, a menos que você não queira considerar o
trecho que faz rastreamento de uma operação que falhou como um trecho de erro. O
método `RecordError` **não** define automaticamente o estado de um trecho ao ser
invocado.

> ### Context e Propagators {#propagators-and-context}

Reviewer: ```suggestion
### Contexto e Propagadores {#propagators-and-context}
```
Aqui não deveria ser em PT-BR?  
https://opentelemetry.io/pt/docs/concepts/context-propagation/

dos panos e gerenciar a serialização de contexto.

## Métricas {#metrics}

Para começar a produzir [métricas](/docs/concepts/signals/metrics), você
precisará ter um `MeterProvider` inicializado que permita a criação de um
`Meter`. Os Meters permitem que você crie instrumentos que podem ser utilizados
para gerar diferentes tipos de métricas. O OpenTelemetry Go suporta atualmente
os seguintes instrumentos:

- **Counter**: Um valor que acumula com o tempo -- você pode imaginar isso como
  um odômetro de um carro; é um valor que só cresce.
- **Asynchronous Counter**: Assim como o **Counter**, porém é coletado uma vez a
  cada exportação. Pode ser usado em casos onde você não tenha acesso aos
  incrementos contínuos, mas apenas ao valor agregado.
- **Histogram**: Uma agregação de valores, tal como latências de requisições. Um
  histograma é uma boa escolha se você está interessado em valores de
  estatísticas. Por exemplo: Quantas requisições estão levando menos de 1s?
- **Gauge**: Mede o valor atual no momento da leitura. Um exemplo seria um
  medidor de tanque de combustível de um veículo. Gauges são assíncronos.

> - **UpDownCounter**: Um valor que acumula com o tempo, mas também pode cair. Um

Reviewer: ```suggestion
- **UpDownCounter**: Um valor que acumula com o tempo, mas também pode reduzir. Um
```
Talvez reduzir fique melhor. O que você acha?

Author: Fica mais explicativo! Boa. Valeu!

  exemplo seria o tamanho de uma fila, este valor irá aumentar e diminuir
  conforme o número de itens que estão entrando ou saindo desta fila.
- **Asynchronous UpDownCounter**: Assim como o **UpDownCounter**, porém é
  coletado uma vez a cada exportação. Pode ser usado em casos onde você não
  tenha acesso às mudanças contínuas, mas apenas ao valor agregado (ex., atual
  tamanho da fila).

Para mais informações sobre instrumentos síncronos, assíncronos, e entender qual
dos tipos melhor se encaixa no seu caso de uso, consulte as
[Diretrizes Suplementares](/docs/specs/otel/metrics/supplementary-guidelines/).

Caso um `MeterProvider` não seja criado, tanto por uma biblioteca de
instrumentação ou manualmente, a API de Métricas do OpenTelemetry usará uma

> implementação no-op e não irá gerar dados.

Reviewer: ```suggestion
implementação sem operação e não irá gerar dados.
```
Talvez facilite a compreensão pra quem não manja muito de desenvolvimento.

A seguir, você poderá encontrar uma documentação mais detalhada para os pacotes:

- Metrics API: [`go.opentelemetry.io/otel/metric`][]
- Metrics SDK: [`go.opentelemetry.io/otel/sdk/metric`][]

> ### Inicializar Métricas {#initialize-metrics}

Reviewer: Parei a revisão aqui. Continuo a revisão em outro momento.

{{% alert color="info" %}} Caso você esteja instrumentando uma biblioteca, pule
esta etapa. {{% /alert %}}

Para habilitar [métricas](/docs/concepts/signals/metrics/) em sua aplicação,
você precisará de um
[`MeterProvider`](/docs/concepts/signals/metrics/#meter-provider) inicializado,
que permitirá que você crie um [`Meter`](/docs/concepts/signals/metrics/#meter).

Caso um `MeterProvider` não seja inicializado, a API de métricas do

> OpenTelemetry irá utilizar uma implementação no-op e não irá gerar dados de

Reviewer: ```suggestion
OpenTelemetry irá utilizar uma implementação sem operação e não irá gerar dados de
```
Acredito que facilita o entendimento.

não-aditivos.

Por exemplo, veja como é possível relatar a velocidade atual de um ventilador de
CPU:

```go
import (
	"net/http"

	"go.opentelemetry.io/otel/metric"
)

var fanSpeedSubscription chan int64

func init() {
	speedGauge, err := meter.Int64Gauge(
		"cpu.fan.speed",
		metric.WithDescription("Velocidade atual de um ventilador de CPU"),
		metric.WithUnit("RPM"),
	)
	if err != nil {
		panic(err)
	}

	getCPUFanSpeed := func() int64 {
		// Vamos gerar um valor aleatório para propósito de demonstração.
		// Em uma aplicação do mundo real, substitua esta implementação para obter o real valor do ventilador de CPU.
		return int64(1500 + rand.Intn(1000))
	}

	fanSpeedSubscription = make(chan int64, 1)
	go func() {
		defer close(fanSpeedSubscription)

		for idx := 0; idx < 5; idx++ {
			// Gauges síncronos são utilizados quando o ciclo de medição está
			// relacionado a uma mudança externa.
			time.Sleep(time.Duration(rand.Intn(3)) * time.Second)
			fanSpeed := getCPUFanSpeed()
			fanSpeedSubscription <- fanSpeed
		}
	}()
}

func recordFanSpeed() {
	ctx := context.Background()
	for fanSpeed := range fanSpeedSubscription {
		speedGauge.Record(ctx, fanSpeed)
	}
}
```

> ### Using Histograms {#using-histograms}

Reviewer: ```suggestion
### Utilizando Histograms {#using-histograms}
```

> Histogramas são utilizados para medir a distribuição de valores ao longo do

Reviewer: ```suggestion
Histograms são utilizados para medir a distribuição de valores ao longo do
```
Olhando a doc. não traduzimos os tipos de métricas:
https://opentelemetry.io/pt/docs/concepts/signals/metrics/

Reviewer 2: Eu acho que faz sentido traduzir não?

Author: @emdneto @EzzioMoreira Na minha cabeça tb faz sentido traduzir, não é uma mudança tão brusca e acaba sendo mais fácil de associar ao tipo de métrica que é gerada.

Caso validem que seja um caso de tradução, posso subir estas alterações na página de métricas logo em seguida. O que acham?

Author: Feito! Fiz o commit ajustando as traduções, seguidas dos termos originais. :)

// Certifique-se de desfazer o registro da métrica `metric.Registration` antes de encerrar a conexão com o banco de dados.
func registerDBMetrics(db *sql.DB, meter metric.Meter, poolName string) (metric.Registration, error) {
	max, err := meter.Int64ObservableUpDownCounter(
		"db.client.connections.max",
		metric.WithDescription("O número máximo de conexões abertas permitidas."),
		metric.WithUnit("{connection}"),
	)
	if err != nil {
		return nil, err
	}

	waitTime, err := meter.Int64ObservableUpDownCounter(
		"db.client.connections.wait_time",
		metric.WithDescription("O tempo que levou para obter uma conexão aberta no pool."),
		metric.WithUnit("ms"),
	)
	if err != nil {
		return nil, err
	}

	reg, err := meter.RegisterCallback(
		func(_ context.Context, o metric.Observer) error {
			stats := db.Stats()
			o.ObserveInt64(max, int64(stats.MaxOpenConnections))
			o.ObserveInt64(waitTime, int64(stats.WaitDuration))
			return nil
		},
		max,
		waitTime,
	)
	if err != nil {
		return nil, err
	}
	return reg, nil
}
```

### Utilizando Gauges Observáveis (Async) {#using-observable-async-gauges}

Gauges observáveis devem ser utilizados para medir valores não-aditivos.

> Por exemplo, veja como é possível reportar o uso de memória dos objetos do heap

Reviewer: ```suggestion
Por exemplo, veja como é possível reportar o uso de memória dos objetos do _heap_
```

var view metric.View = func(i metric.Instrument) (metric.Stream, bool) {
	// Em uma função de View personalizada, é necessário copiar de maneira explícita
	// o nome, descrição e unidade.
	s := metric.Stream{Name: i.Name, Description: i.Description, Unit: i.Unit}
	// Qualquer instrumento que não tenha um sufixo de unidade de medida definido, mas tenha
	// uma unidade dimensional definida, terá seu nome atualizado com um sufixo de unidade de medida.
	if re.MatchString(i.Name) {
		return s, false
	}
	switch i.Unit {
	case "ms":
		s.Name += ".ms"
	case "By":
		s.Name += ".byte"
	default:
		return s, false
	}
	return s, true
}

meterProvider := metric.NewMeterProvider(
	metric.WithView(view),
)
```

## Logs

Logs são distintos de métricas e rastros, pois **não há uma API de logs do
OpenTelemetry voltada para o usuário**. Em vez disso, existem ferramentas que
integram pacotes de logs populares existentes (como slog, logrus, zap, logr) aos
logs do ecossistema OpenTelemetry. Para a justificativa por trás dessa decisão
de design, consulte a [Especificação de Logs](/docs/specs/otel/logs/).

Os dois _workflows_ discutidos abaixo atendem a diferentes requisitos de
aplicação.

> ### Direct-to-Collector

Reviewer: ```suggestion
### Direto para o Collector {direct-to-collector}
```

Reviewer: Acredito que deveriamos traduzir.

**Estado**: [Experimental](/docs/specs/otel/document-status/)

Dentro do _workflow_ Direct-to-collector, os logs são emitidos diretamente da
aplicação para o Collector utilizando um protocolo de rede (por exemplo, OTLP).
Este _workflow_ é simples de configurar, pois não requer componentes adicionais
de encaminhamento de logs, permitindo que uma aplicação emita logs estruturados
e que estejam nos conformes do [modelo de dados de logs][log data model]. No
entanto, o _overhead_ necessário para que as aplicações enfileirem e exportem
logs para um local de rede pode não ser adequado para todas as aplicações.

Para utilizar este _workflow_:

- Configurar o [SDK de Logs](#logs-sdk) do OpenTelemetry para exportar registros
  de logs para o destino desejado (o [OpenTelemetry
  Collector][opentelemetry collector] ou outro).

> - Utilizar uma [Ponte de Logs](#log-bridge) apropriada.

Reviewer: ```suggestion
- Utilizar uma [Log Bridge](#log-bridge) apropriada.
```
Como se trata de um recurso do projeto otel, não deveriamos traduzir.

Author: Hmm, atualmente estamos traduzindo o termo na página de conceitos: https://opentelemetry.io/pt/docs/concepts/signals/logs/#log-appender--bridge

Qual seria o melhor caminho?

Reviewer 2: Se nos conceitos estamos traduzindo, acredito que podemos traduzir aqui. @EzzioMoreira 
Talvez apenas colocar o termo em inglês (bridge) entre parênteses pra fazer a referência? Pq ponte/bridge são conceitos conhecidos no mundo de software

Author: Feito! 

Ajustei as traduções para utilizarmos Log Bridge

#### SDK de Logs {#logs-sdk}

A SDK de Logs define como os logs são processados ao utilizar o _workflow_

> [Direct-to-Collector](#direct-to-collector). Nenhuma SDK de logs é necessária ao

Reviewer: ```suggestion
[Diretamente para o Collector](#direct-to-collector). Nenhuma SDK de logs é necessária ao
```
Se aceitarem traduzir `Direct-to-Collector`, precisa mudar aqui.

utilizar o _workflow_ de [encaminhamento de logs](#via-file-or-stdout).

Uma configuração típica da SDK de logs instala um Processor de logs em lote com
um Exporter OTLP.

Para habilitar [logs](/docs/concepts/signals/logs/) em sua aplicação, você
precisará ter um
[`LoggerProvider`](/docs/concepts/signals/logs/#logger-provider) inicializado,

> que permitirá que você utilize uma [Ponte de Logs](#log-bridge).

Reviewer: ```suggestion
que permitirá que você utilize uma [Log Bridge](#log-bridge).
```

	if err != nil {
		panic(err)
	}

	// Lidamos com a finalização corretamente, evitando leaks.
	defer func() {
		if err := loggerProvider.Shutdown(ctx); err != nil {
			fmt.Println(err)
		}
	}()

	// Registre o LoggerProvider globalmente, permitindo que seja acessado via global.LoggerProvider.
	// A maioria das pontes de logs utilizam o LoggerProvider global como padrão.
	// Caso o LoggerProvider global não esteja definido, será utilizada uma implementação no-op
	// e dados de logs não serão gerados.
	global.SetLoggerProvider(loggerProvider)
}

func newResource() (*resource.Resource, error) {
	return resource.Merge(resource.Default(),
		resource.NewWithAttributes(semconv.SchemaURL,
			semconv.ServiceName("meu-servico"),
			semconv.ServiceVersion("0.1.0"),
		))
}

func newLoggerProvider(ctx context.Context, res *resource.Resource) (*log.LoggerProvider, error) {
	exporter, err := otlploghttp.New(ctx)
	if err != nil {
		return nil, err
	}
	processor := log.NewBatchProcessor(exporter)
	provider := log.NewLoggerProvider(
		log.WithResource(res),
		log.WithProcessor(processor),
	)
	return provider, nil
}
```

Agora que o `LoggerProvider` está configurado, é possível utilizá-lo para

> configurar uma [Ponte de Logs](#log-bridge).

Reviewer: ```suggestion
configurar uma [Log Bridge](#log-bridge).
```

> #### Ponte de Logs {#log-bridge}

Reviewer: ```suggestion
#### Log Bridge {#log-bridge}
```

> Uma ponte de logs é um componente que conecta logs de um pacote existente ao

Reviewer: ```suggestion
Uma Log Bridge é um componente que conecta logs de um pacote existente ao
```

> [SDK de Logs](#logs-sdk) do OpenTelemetry, utilizando a [API de Ponte de

Reviewer: ```suggestion
[SDK de Logs](#logs-sdk) do OpenTelemetry, utilizando a [Logs Bridge API]
```

> Logs][logs bridge API].

Reviewer: ```suggestion
```

Reviewer: Essa parte foi movida para linha de cima.

> Uma lista completa contendo as pontes de logs disponíveis pode ser encontrada no

Reviewer: ```suggestion
Uma lista completa contendo os log bridges disponíveis pode ser encontrada no
```

[registro do OpenTelemetry](/ecosystem/registry/?language=go&component=log-bridge).

Cada pacote de ponte de logs deve possuir uma documentação contendo as suas
instruções de instalação e configuração.

### Através de arquivos ou stdout {#via-file-or-stdout}

No _workflow_ utilizado em arquivos ou _stdout_, os logs são gravados em
arquivos ou na saída padrão da aplicação. Outro componente (por exemplo,

> FluentBit) é responsável por ler/seguir os logs, convertê-los para um formato

Reviewer: ```suggestion
FluentBit) é responsável por ler/acompanhar os logs, convertê-los para um formato
```
Acha que acompanhar ficar melhor? No sentido de observar linhas de logs geradas.

Author: Sim, fica melhor! Não me veio à cabeça no momento um termo equivalente, mas fica muito melhor. Gracias!

mais estruturado e encaminhá-los para um destino, como o Collector. Este
_workflow_ pode ser preferível em situações onde os requisitos da aplicação não

> permitem a sobrecarga adicional do [Direct-to-Collector](#direct-to-collector).

Reviewer: ```suggestion
permitem a sobrecarga adicional do [Diretamente para o Collector](#direct-to-collector).
```

---
title: Instrumentação
aliases:
  - manual
  - manual_instrumentation
weight: 30
description: Instrumentação manual para OpenTelemetry Go
default_lang_commit: 1c6697de9c4d67fb72231354d5d9c6cdcfdfa64b
# prettier-ignore
cSpell:ignore: fatalf logr logrus otelslog otlplog otlploghttp sdktrace sighup updown
---

{{% docs/languages/instrumentation-intro %}}

## Configuração {#setup}

## Rastros {#traces}

### Inicializando um Tracer {#getting-a-tracer}

Para criar trechos, você precisará obter ou inicializar um Tracer primeiro.

> Certifique-se de haver instalado corretamente os seguintes pacotes Go:

Reviewer: ```suggestion
Certifique-se que os seguintes pacotes Go foram instalados corretamente:
```

	// invoque a função filha e inicie um trecho aninhado dentro dela
	childFunction(ctx)

	// insira mais coisas no código - ao finalizar esta função, o trecho declarado em 'parentSpan' será finalizado.
}

func childFunction(ctx context.Context) {
	// Inicialize um trecho para rastrear a chamada `childFunction()`
	// este é um trecho aninhado, cujo parente foi declarado no método anterior em `parentSpan`
	ctx, childSpan := tracer.Start(ctx, "child")
	defer childSpan.End()

	// insira algum código aqui - ao finalizar esta função, o trecho declarado em 'childSpan' será finalizado.
}
```

Uma vez que o trecho é finalizado, ele se torna imutável e não pode mais ser
modificado.

### Atributos de Trecho {#span-attributes}

Os atributos são pares de chaves e valor aplicados como metadados aos seus
trechos e são úteis para agregar, filtrar e agrupar rastros. Os atributos podem
ser adicionados durante a criação de um trecho ou a qualquer momento durante seu
ciclo de vida, antes que ele seja concluído.

```go
// definindo atributos na criação...
ctx, span = tracer.Start(ctx, "atributosNaCriacao", trace.WithAttributes(attribute.String("hello", "world")))
// ... e após a criação

> span.SetAttributes(attribute.Bool("ehVerdadeiro", true), attribute.String("atributoDeTexto", "oi!"))

Author: Boa! Faz sentido, heheh. Obrigado pelo comentário!

```

As chaves dos atributos também podem ser pré-computadas:

```go
var myKey = attribute.Key("meuAtributoLegal")
span.SetAttributes(myKey.String("um valor em texto"))
```

#### Atributos Semânticos {#semantic-attributes}

Os Atributos Semânticos são atributos definidos pela [Especificação do
OpenTelemetry][opentelemetry specification] para fornecer um conjunto comum de
chaves de atributos entre várias linguagens, frameworks e ambientes de execução.

> Eles representam conceitos como métodos HTTP, códigos de estado, user agents e

Reviewer: ```suggestion
Esses atributos representam conceitos como métodos HTTP, códigos de estado, user agents e
```
Estamos usando pouco pronomes, talvez assim fique melhor.

outros. Estes atributos estão disponíveis no pacote
`go.opentelemetry.io/otel/semconv/v1.26.0`.

Para mais detalhes, consulte as [Convenções Semânticas de
Rastros][trace semantic conventions].

### Eventos {#events}

Um evento é uma mensagem legível para humanos em um trecho que representa "algo
acontecendo" durante a sua duração. Por exemplo, imagine uma função que requer
acesso exclusivo a um recurso que está sob um mutex. Um evento poderia ser
criado em dois pontos: um quando tentamos obter acesso ao recurso e outro quando
adquirimos o mutex.

```go
span.AddEvent("Obtendo o lock")
mutex.Lock()
span.AddEvent("Lock efetuado, realizando o trabalho...")
// do stuff
span.AddEvent("Realizando unlocking")
mutex.Unlock()
```

Uma característica útil dos eventos é que seus timestamps são exibidos como

> offsets a partir do início do span, permitindo ver facilmente quanto tempo se

Reviewer: ```suggestion
offsets a partir do início do trecho, permitindo ver facilmente quanto tempo se
```

span.SetAttributes(attribute.Bool("isTrue", true), attribute.String("atributoDeTexto", "oi!"))
```

As chaves dos atributos também podem ser pré-computadas:

```go
var myKey = attribute.Key("meuAtributoLegal")
span.SetAttributes(myKey.String("um valor em texto"))
```

#### Atributos Semânticos {#semantic-attributes}

Os Atributos Semânticos são atributos definidos pela [Especificação do
OpenTelemetry][opentelemetry specification] para fornecer um conjunto comum de
chaves de atributos entre várias linguagens, frameworks e ambientes de execução.
Esses atributos representam conceitos como métodos HTTP, códigos de estado, user
agents e outros. Estes atributos estão disponíveis no pacote
`go.opentelemetry.io/otel/semconv/v1.26.0`.

Para mais detalhes, consulte as [Convenções Semânticas de
Rastros][trace semantic conventions].

### Eventos {#events}

Um evento é uma mensagem legível para humanos em um trecho que representa "algo
acontecendo" durante a sua duração. Por exemplo, imagine uma função que requer
acesso exclusivo a um recurso que está sob um mutex. Um evento poderia ser
criado em dois pontos: um quando tentamos obter acesso ao recurso e outro quando
adquirimos o mutex.

```go

> span.AddEvent("Obtendo o lock")

Reviewer 3: ```suggestion
span.AddEvent("Obtendo o bloqueio")
```

mutex.Lock()

> span.AddEvent("Lock efetuado, realizando o trabalho...")

Reviewer 3: ```suggestion
span.AddEvent("Bloqueio efetuado, realizando ações...")
```

// fazendo algo
span.AddEvent("Realizando o desbloqueio")
mutex.Unlock()
```

Uma característica útil dos eventos é que seus timestamps são exibidos como
intervalos a partir do início do trecho, permitindo ver facilmente quanto tempo
se passou entre cada um.

> Os Eventos também podem incluir seus próprios atributos -

Reviewer 3: ```suggestion
Os eventos também podem incluir seus próprios atributos -
```

```go

> span.AddEvent("Espera cancelada devido sinal externo", trace.WithAttributes(attribute.Int("pid", 4328), attribute.String("signal", "SIGHUP")))

Reviewer 3: ```suggestion
span.AddEvent("Espera cancelada devido a um sinal externo", trace.WithAttributes(attribute.Int("pid", 4328), attribute.String("signal", "SIGHUP")))
```

```

> ### Definir status do Trecho {#set-span-status}

Reviewer 3: ```suggestion
### Definir status do trecho {#set-span-status}
```

{{% pt/docs/languages/span-status-preamble %}}

```go
import (
	// ...
	"go.opentelemetry.io/otel/codes"
	// ...
)

// ...

result, err := operacaoQueDeveriaFalhar()
if err != nil {
	span.SetStatus(codes.Error, "operacaoQueDeveriaFalhar falhou")
}
```

### Capturar erros {#record-errors}

Caso você tenha uma operação que falhou e deseja capturar o erro produzido, você

> pode registrar este erro.

Reviewer 3: ```suggestion
pode salvar este erro.
```

```go
import (
	// ...
	"go.opentelemetry.io/otel/codes"
	// ...
)

// ...

result, err := operacaoQueDeveriaFalhar()
if err != nil {
	span.SetStatus(codes.Error, "operacaoQueDeveriaFalhar falhou")
	span.RecordError(err)
}
```

É altamente recomendável que você também defina o estado de um trecho como
`Error` ao utilizar `RecordError`, a menos que você não queira considerar o
trecho que faz rastreamento de uma operação que falhou como um trecho de erro. O
método `RecordError` **não** define automaticamente o estado de um trecho ao ser
invocado.

### Contexto e Propagadores {#propagators-and-context}

> Os Rastros podem se estender além de um único processo. Isso requer a

Reviewer 3: ```suggestion
Os rastros podem se estender além de um único processo. Isso requer a
```

_propagação de contexto_, um mecanismo onde os identificadores de um rastro são
enviados para processos remotos.

Para propagar o contexto de um rastro pela rede, um propagador deve ser
registrado com a API do OpenTelemetry.

```go
import (
  "go.opentelemetry.io/otel"
  "go.opentelemetry.io/otel/propagation"
)
...
otel.SetTextMapPropagator(propagation.TraceContext{})
```

> O OpenTelemetry também suporta cabeçalhos no formato B3, para compatibilidade
> com sistemas de rastreamento (`go.opentelemetry.io/contrib/propagators/b3`)
> que não suportam o padrão W3C TraceContext.

Após configurar a propagação de contexto, você provavelmente vai querer utilizar
a instrumentação automática para lidar com todo o trabalho que acontece debaixo

> dos panos e gerenciar a serialização de contexto.

Reviewer 3: ```suggestion
dos panos para gerenciar a serialização de contexto.
```

## Métricas {#metrics}

Para começar a produzir [métricas](/docs/concepts/signals/metrics), você
precisará ter um `MeterProvider` inicializado que permita a criação de um
`Meter`. Os Meters permitem que você crie instrumentos que podem ser utilizados
para gerar diferentes tipos de métricas. O OpenTelemetry Go suporta atualmente
os seguintes instrumentos:

> - **Counter**: Um valor que acumula com o tempo -- você pode imaginar isso como

Author: Ahhh, sim!

O que aconteceu aqui foi que acabei buscando o texto que está em [Conceitos -> Sinais -> Métricas](https://opentelemetry.io/pt/docs/concepts/signals/metrics/) (que por coincidência, a tradução inicial esteve comigo, rs) e acabei mantendo o mesmo padrão. Não cheguei a me atentar que os tipos de Gauge não estão iguais -- inclusive, vale a pena abrir um novo PR para atualizarmos a seção `Metric Instruments`?

> - **Counter**: Um valor que acumula com o tempo -- você pode imaginar isso como

Reviewer 3: Acho que seria legal atualizar a seção `Metric Instruments` com qualquer informação que tenha faltado sim! Mas seria bom atualizar tanto inglês quanto português. Se você quiser, você pode atualizar no português e eu faço no inglês seguindo o que você fizer pra ficar alinhado, só me avisar

> - **Counter**: Um valor que acumula com o tempo -- você pode imaginar isso como

Reviewer 2: Oq acham de criarmos uma issue pra trackear isso? só pra não ser perder

> - **Counter**: Um valor que acumula com o tempo -- você pode imaginar isso como

Reviewer 3: Acho melhor manter como no original mesmo, e uma boa idéia criar a issue pra adicionar exemplos.
Mas acho que você esqueceu de commitar as mudanças pra voltar pro original, porque essa versão está diferente da versão em inglês.
Pra deixar claro, em inglês está:

```
- Counter, a synchronous instrument that supports non-negative increments
- Asynchronous Counter, an asynchronous instrument which supports non-negative
  increments
- Histogram, a synchronous instrument that supports arbitrary values that are
  statistically meaningful, such as histograms, summaries, or percentile
- Synchronous Gauge, a synchronous instrument that supports non-additive values,
  such as room temperature.
- Asynchronous Gauge, an asynchronous instrument that supports non-additive
  values, such as room temperature
- UpDownCounter, a synchronous instrument that supports increments and
  decrements, such as the number of active requests
- Asynchronous UpDownCounter, an asynchronous instrument that supports
  increments and decrements
  ```

  um odômetro de um carro; é um valor que só cresce.
- **Asynchronous Counter**: Assim como o **Counter**, porém é coletado uma vez a
  cada exportação. Pode ser usado em casos onde você não tenha acesso aos
  incrementos contínuos, mas apenas ao valor agregado.

> - **Histogram**: Uma agregação de valores, tal como latências de requisições. Um

Author: ✅

> - **Histogram**: Uma agregação de valores, tal como latências de requisições. Um

Reviewer 3: Parece estar faltando mesmo

  histograma é uma boa escolha se você está interessado em valores de
  estatísticas. Por exemplo: Quantas requisições estão levando menos de 1s?

> - **Gauge**: Mede o valor atual no momento da leitura. Um exemplo seria um

Author: ✅

> - **Gauge**: Mede o valor atual no momento da leitura. Um exemplo seria um

Reviewer 2: falta corrigir aqui ainda @maryliag @vitorvasc ?

> - **Gauge**: Mede o valor atual no momento da leitura. Um exemplo seria um

Reviewer 3: está faltando mesmo

  go.opentelemetry.io/otel/sdk \
  go.opentelemetry.io/otel/sdk/metric
```

Em seguida, inicialize um `Resource`, `Metrics Exporter` e um `Meter Provider`:

```go
package main

import (
	"context"
	"log"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/stdout/stdoutmetric"
	"go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
)

func main() {
	// Crie um 'Resource'.
	res, err := newResource()
	if err != nil {
		panic(err)
	}

	// Inicialize um Meter Provider.
	// Você poderá passar a instância diretamente para o seu código instrumentado, caso
	// o mesmo aceite uma instância do MeterProvider.
	meterProvider, err := newMeterProvider(res)
	if err != nil {
		panic(err)
	}

	// Lidamos com a finalização corretamente, evitando vazamentos.
	defer func() {
		if err := meterProvider.Shutdown(context.Background()); err != nil {
			log.Println(err)
		}
	}()

	// Registre o MeterProvider globalmente, permitindo a utilização via otel.Meter
	// e o acesso utilizando otel.GetMeterProvider.
	// A maioria das bibliotecas de instrumentação utilizam o MeterProvider global como padrão.

> 	// Caso o MeterProvider global não esteja definido, será utilizada uma implementação sem oepração

Reviewer 3: ops... escrevi errado na sugestão... sorry

```suggestion
	// Caso o MeterProvider global não esteja definido, será utilizada uma implementação sem operação
```

  deseja que o processo em execução aguarde pela medição.
- Quando medições precisam acontecer em frequências não-relacionadas ao tempo de
  execução da aplicação (ou seja, não podem ser medidas com precisão ao serem
  vinculadas ao ciclo de vida de uma operação).
- Quando não há um registro de data/hora conhecidos para um valor de medição.

Em casos como estes, muitas vezes é melhor realizar a medição de um valor
cumulativo diretamente, em vez de agregar uma série de deltas no
pós-processamento (como ocorre no exemplo síncrono).

### Utilizando Contadores (Counters) {#using-counters}

Contadores (Counters) podem ser utilizados para medir valores incrementais e
não-negativos.

Por exemplo, aqui está como seria possível reportar o número de chamadas em um
servidor HTTP:

```go
import (
	"net/http"

	"go.opentelemetry.io/otel/metric"
)

func init() {
	apiCounter, err := meter.Int64Counter(
		"api.counter",
		metric.WithDescription("Número de chamadas na API."),
		metric.WithUnit("{call}"),
	)
	if err != nil {
		panic(err)
	}
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		apiCounter.Add(r.Context(), 1)

		// implementação da chamada da API
	})
}
```

> ### Utilizando Contadores UpDown (Updown Counters) {#using-updown-counters}

Reviewer 3: ```suggestion
### Utilizando Contadores UpDown (UpDown Counters) {#using-updown-counters}
```

			"Uso de memória dos objetos alocados no heap.",
		),
		metric.WithUnit("By"),
		metric.WithInt64Callback(func(_ context.Context, o metric.Int64Observer) error {
			var m runtime.MemStats
			runtime.ReadMemStats(&m)
			o.Observe(int64(m.HeapAlloc))
			return nil
		}),
	); err != nil {
		panic(err)
	}
}
```

### Adicionando atributos {#adding-attributes}

É possível adicionar atributos utilizando as opções
[`WithAttributeSet`](https://pkg.go.dev/go.opentelemetry.io/otel/metric#WithAttributeSet)
ou
[`WithAttributes`](https://pkg.go.dev/go.opentelemetry.io/otel/metric#WithAttributes).

```go
import (
	"net/http"

	"go.opentelemetry.io/otel/metric"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
)

func init() {
	apiCounter, err := meter.Int64UpDownCounter(
		"api.finished.counter",
		metric.WithDescription("Número de chamadas finalizadas na API."),
		metric.WithUnit("{call}"),
	)
	if err != nil {
		panic(err)
	}
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		// insira algum código na chamada da API e defina o status HTTP de resposta

		apiCounter.Add(r.Context(), 1,
			metric.WithAttributes(semconv.HTTPResponseStatusCode(statusCode)))
	})
}
```

> ### Registering Views {#registering-views}

Reviewer 3: ```suggestion
### Registrando Views {#registering-views}
```

> Uma _view_ oferece aos usuários a flexibilidade de personalizar a emissão das

Reviewer 3: ```suggestion
Uma _view_ oferece aos usuários do SDK a flexibilidade de personalizar a emissão das
```

> métricas fornecidas pelo SDK. Você pode personalizar quais instrumentos de

Reviewer 3: ```suggestion
métricas pelo SDK. Você pode personalizar quais instrumentos de
```

não são só métricas fornecidas pelo SDK

É possível utilizar o método
[`NewView`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric#NewView) para
criar uma _view_ e registrá-la utilizando a opção
[`WithView`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric#WithView).

Por exemplo, veja como é possível criar uma _view_ que renomeia o instrumento
`latency` da versão `v0.34.0` da biblioteca de instrumentação `http` para
`request.latency`:

```go
view := metric.NewView(metric.Instrument{
	Name: "latency",
	Scope: instrumentation.Scope{
		Name:    "http",
		Version: "0.34.0",
	},
}, metric.Stream{Name: "request.latency"})

meterProvider := metric.NewMeterProvider(
	metric.WithView(view),
)
```

Em outro exemplo, veja como é possível criar uma _view_ que faz com que o
instrumento `latency` da biblioteca de instrumentação `http` seja reportado como
um histograma exponencial:

```go
view := metric.NewView(
	metric.Instrument{
		Name:  "latency",
		Scope: instrumentation.Scope{Name: "http"},
	},
	metric.Stream{
		Aggregation: metric.AggregationBase2ExponentialHistogram{
			MaxSize:  160,
			MaxScale: 20,
		},
	},
)

meterProvider := metric.NewMeterProvider(
	metric.WithView(view),
)
```

> A SDK filtra métricas e atributos antes de exportar as métricas. Você pode

Reviewer 2: ```suggestion
O SDK filtra métricas e atributos antes de exportar as métricas. Você pode
```

de design, consulte a [Especificação de Logs](/docs/specs/otel/logs/).

Os dois _workflows_ discutidos abaixo atendem a diferentes requisitos de
aplicação.

### Direto para o Collector {#direct-to-collector}

**Estado**: [Experimental](/docs/specs/otel/document-status/)

Dentro do _workflow_ Direct-to-collector, os logs são emitidos diretamente da
aplicação para o Collector utilizando um protocolo de rede (por exemplo, OTLP).
Este _workflow_ é simples de configurar, pois não requer componentes adicionais
de encaminhamento de logs, permitindo que uma aplicação emita logs estruturados
e que estejam nos conformes do [modelo de dados de logs][log data model]. No
entanto, o _overhead_ necessário para que as aplicações enfileirem e exportem
logs para um local de rede pode não ser adequado para todas as aplicações.

Para utilizar este _workflow_:

- Configurar o [SDK de Logs](#logs-sdk) do OpenTelemetry para exportar registros
  de logs para o destino desejado (o [OpenTelemetry
  Collector][opentelemetry collector] ou outro).
- Utilizar uma [Ponte de Logs](#log-bridge) apropriada.

#### SDK de Logs {#logs-sdk}

> A SDK de Logs define como os logs são processados ao utilizar o _workflow_
> [Diretamente para o Collector](#direct-to-collector). Nenhuma SDK de logs é
> necessária ao utilizar o _workflow_ de
> [encaminhamento de logs](#via-file-or-stdout).

Reviewer 2: ```suggestion
A SDK de Logs define como os logs são processados ao utilizar o _workflow_
[Diretamente para o Collector](#direct-to-collector). Nenhum SDK de logs é
necessário ao utilizar o _workflow_ de
[encaminhamento de logs](#via-file-or-stdout).
```

> Uma configuração típica da SDK de logs instala um Processor de logs em lote com

Reviewer 2: ```suggestion
Uma configuração típica do SDK de logs instala um Processor de logs em lote com
```

a inicialização do SDK utilizando os seguintes pacotes:

- [`go.opentelemetry.io/otel`][]
- [`go.opentelemetry.io/otel/sdk/log`][]
- [`go.opentelemetry.io/otel/sdk/resource`][]
- [`go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploghttp`][]

Certifique-se que os seguintes pacotes Go foram instalados corretamente:

```sh
go get go.opentelemetry.io/otel \
  go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploghttp \
  go.opentelemetry.io/otel/sdk \
  go.opentelemetry.io/otel/sdk/log
```

Em seguida, inicialize o `LoggerProvider`:

```go
package main

import (
	"context"
	"fmt"

	"go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploghttp"
	"go.opentelemetry.io/otel/log/global"
	"go.opentelemetry.io/otel/sdk/log"
	"go.opentelemetry.io/otel/sdk/resource"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
)

func main() {
	ctx := context.Background()

	// Crie um 'Resource'.
	res, err := newResource()
	if err != nil {
		panic(err)
	}

	// Crie um 'LoggerProvider'.
	// Você pode passar essa instância diretamente ao criar pontes.
	loggerProvider, err := newLoggerProvider(ctx, res)
	if err != nil {
		panic(err)
	}

> 	// Lidamos com a finalização corretamente, evitando leaks.

Reviewer 3: ```suggestion
	// Lidamos com a finalização corretamente, evitando vazamentos.
```

	defer func() {
		if err := loggerProvider.Shutdown(ctx); err != nil {
			fmt.Println(err)
		}
	}()

	// Registre o LoggerProvider globalmente, permitindo que seja acessado via global.LoggerProvider.
	// A maioria das pontes de logs utilizam o LoggerProvider global como padrão.

> 	// Caso o LoggerProvider global não esteja definido, será utilizada uma implementação no-op

Reviewer 3: ```suggestion
	// Caso o LoggerProvider global não esteja definido, será utilizada uma implementação sem operação
```

	global.SetLoggerProvider(loggerProvider)
}

func newResource() (*resource.Resource, error) {
	return resource.Merge(resource.Default(),
		resource.NewWithAttributes(semconv.SchemaURL,
			semconv.ServiceName("meu-servico"),
			semconv.ServiceVersion("0.1.0"),
		))
}

func newLoggerProvider(ctx context.Context, res *resource.Resource) (*log.LoggerProvider, error) {
	exporter, err := otlploghttp.New(ctx)
	if err != nil {
		return nil, err
	}
	processor := log.NewBatchProcessor(exporter)
	provider := log.NewLoggerProvider(
		log.WithResource(res),
		log.WithProcessor(processor),
	)
	return provider, nil
}
```

Agora que o `LoggerProvider` está configurado, é possível utilizá-lo para
configurar uma [Ponte de Logs](#log-bridge).

#### Ponte de Logs {#log-bridge}

Uma ponte de logs é um componente que conecta logs de um pacote existente ao
[SDK de Logs](#logs-sdk) do OpenTelemetry, utilizando a [API de Ponte de
Logs][logs bridge API].

Uma lista completa contendo as pontes de logs disponíveis pode ser encontrada no
[registro do OpenTelemetry](/ecosystem/registry/?language=go&component=log-bridge).

Cada pacote de ponte de logs deve possuir uma documentação contendo as suas

> instruções de instalação e configuração.

Reviewer 3: ```suggestion
instruções de instalação e configuração, e um exemplo.
```

### Através de arquivos ou stdout {#via-file-or-stdout}

No _workflow_ utilizado em arquivos ou _stdout_, os logs são gravados em
arquivos ou na saída padrão da aplicação. Outro componente (por exemplo,
FluentBit) é responsável por ler/acompanhar os logs, convertê-los para um
formato mais estruturado e encaminhá-los para um destino, como o Collector. Este
_workflow_ pode ser preferível em situações onde os requisitos da aplicação não
permitem a sobrecarga adicional do

> [Diretamente para o Collector](#direct-to-collector). No entanto, é requisito

Reviewer 3: ```suggestion
[Direto para o Collector](#direct-to-collector). No entanto, é requisito
```
